pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,343 @@
+            self._current_action = next(self._schedule)
+            self.forward_schedule.append(self._current_action)
+
+    def end_taping(self):
+        """Process the end of the forward execution."""
+        current_timestep = self.tape.latest_timestep
+        while self.mode != Mode.EVALUATED:
+            self.end_timestep(current_timestep)
+            current_timestep += 1
+
+    @singledispatchmethod
+    def process_taping(self, cp_action, timestep):
+        """A single-dispatch generic function.
+        The process taping is used while the forward model is taped.
+
+        Parameters
+        ----------
+        cp_action : checkpoint_schedules.CheckpointAction
+            A checkpoint action obtained from the `checkpoint_schedules`.
+            In process taping, the possible actions are `Forward` and `EndForward`.
+            These actions are used to manage the execution of the forward model,
+            store checkpoint data, and inform the end of the forward execution.
+        timestep : int
+            The current timestep.
+
+        Returns
+        -------
+        bool
+            Returns `True` if the timestep is in the `checkpoint_schedules` action.
+            For example, if the `checkpoint_schedules` action is
+            `Forward(2, 4, True, False, StorageType.DISK)`, then timestep `3` is considered within the action;
+            otherwise, it returns `False`.
+
+        Raises
+        ------
+        CheckpointError
+            If the checkpoint action is not supported.
+
+        Notes
+        -----
+        To have more information about the `checkpoint_schedules`,
+        please refer to the
+        `documentation <https://www.firedrakeproject.org/checkpoint_schedules/>`_.
+        Detailed descriptions of the actions used in the process taping can be found at the following links:
+        `Forward
+        <https://www.firedrakeproject.org/checkpoint_schedules/checkpoint_schedules.html#checkpoint_schedules.schedule.Forward>`_
+        and `End_Forward
+        <https://www.firedrakeproject.org/checkpoint_schedules/checkpoint_schedules.html#checkpoint_schedules.schedule.EndForward>`_.
+        """
+
+        raise CheckpointError(f"Unable to process {cp_action} while taping.")
+
+    @process_taping.register(Forward)
+    def _(self, cp_action, timestep):
+        if timestep < (cp_action.n0):
+            raise CheckpointError(
+                "Timestep is before start of Forward action."
+            )
+
+        self.tape._eagerly_checkpoint_outputs = cp_action.write_adj_deps
+
+        if timestep > cp_action.n0:
+            if cp_action.write_ics and timestep == (cp_action.n0 + 1):
+                # Stores the checkpoint data in RAM
+                # This data will be used to restart the forward model
+                # from the step `n0` in the reverse computations.
+                self.tape.timesteps[cp_action.n0].checkpoint()
```

Comment:
Can I check about the case were forward data computed in step $n$ is used for forward calculations in steps $\ge n + 2$.

e.g. for the Firedrake code
```
from firedrake import *
from firedrake.adjoint import continue_annotation

mesh = UnitIntervalMesh(1)
space = FunctionSpace(mesh, "DG", 0)
U = [Function(space).assign(n) for n in range(N)]

continue_annotation()
v = Function(space)
for u in get_working_tape().timestepper(U):
   v = Function(space).assign(v + u)
```
the elements of `U` all have `last_use = -1`, and are each used in exactly one step. A checkpoint for a forward restart at step the start of step `n0` advancing to the start of step `n1` needs to include `U[n0:n1]`.

Storing the checkpoint after advancing to the start of step `n0 + 1` looks wrong, unless later `U` elements are flagged for checkpointing at step `n0`. Including too many `U` elements uses too much storage.

I might just be missing how this works (I see other relevant code in tape.py), but I suggest adding a test case.

Reply:
A different example, with a structure that might be encountered with an explicit linear multistep discretization:

```
from firedrake import *
from firedrake.adjoint import continue_annotation, get_working_tape

from collections import deque

M, N = 3, 100

mesh = UnitIntervalMesh(1)
space = FunctionSpace(mesh, "DG", 0)
U = deque([Function(space).assign(m) for m in range(M)])

continue_annotation()
v = Function(space)
for _ in get_working_tape().timestepper(range(N)):
   U[0].assign(U[-1])
   U.append(U.popleft())
```